src/wasm/wasm-type.cpp

```diff
@@ -983,4 +945,115 @@ std::ostream& operator<<(std::ostream& os, TypeInfo info) {
   WASM_UNREACHABLE("unexpected kind");
 }
 
+std::ostream& operator<<(std::ostream& os, HeapTypeInfo info) {
+  switch (info.kind) {
+    case HeapTypeInfo::SignatureKind:
+      return os << info.signature;
+    case HeapTypeInfo::StructKind:
+      return os << info.struct_;
+    case HeapTypeInfo::ArrayKind:
+      return os << info.array;
+  }
+  WASM_UNREACHABLE("unexpected kind");
+}
+
 } // namespace wasm
+
+namespace std {
```

Comment:
Any reason we put `hash` functions in two separate chunks? Can we merge them into one chunk?

Reply:
There are a couple unfortunate things going on here.

 1. The declaration of specializations of `std::hash` for `TypeInfo` and `HeapTypeInfo` needs to be done before the declaration of `Store`, otherwise the declaration of `Store::typeIDs` tries to use the default instantiation and fails.

 2. There's no way to declare specializations from a different namespace, so there's no way to avoid leaving the anonymous and `wasm` namespaces and entering `std` just for these two declarations.

There's nothing stopping us from moving the hash definitions from down below up here, but I thought it would still be better to keep them separate despite the extra verbosity.